cyw43/src/bluetooth.rs

```diff
@@ -0,0 +1,165 @@
+use embassy_time::{Duration, Timer};
+use embedded_hal_1::digital::OutputPin;
+
+use crate::bus::Bus;
+use crate::consts::*;
+use crate::{SpiBusCyw43, CHIP};
+
+fn is_aligned(a: u32, x: u32) -> bool {
+    (a & (x - 1)) == 0
+}
+
+fn round_down(x: u32, a: u32) -> u32 {
+    x & !(a - 1)
+}
+
+pub(crate) async fn upload_bluetooth_firmware<PWR: OutputPin, SPI: SpiBusCyw43>(
+    bus: &mut Bus<PWR, SPI>,
+    firmware_offsets: &[(u32, usize)],
+    firmware: &[u8],
+) {
+    // read version
+    let version_length = firmware[0];
+    let _version = &firmware[1..=version_length as usize];
+    // skip version + 1 extra byte as per cybt_shared_bus_driver.c
+    let firmware = &firmware[version_length as usize + 2..];
+    // buffer
+    let mut data_buffer: [u8; 0x100] = [0; 0x100];
+    let mut aligned_data_buffer: [u8; 0x100] = [0; 0x100];
+    // structs
+    let mut pointer = 0;
+    for (index, &(dest_addr, num_fw_bytes)) in firmware_offsets.iter().enumerate() {
```

Comment:
The memory being used here is getting overriden or something... it isn't behaving as expected (0, 1, 2, 3, .. 27) for a 28 element slice